client/websocket_client.go

```diff
@@ -0,0 +1,193 @@
+
+	"github.com/eris-ltd/eris-db/txs"
+	ctypes "github.com/eris-ltd/eris-db/rpc/tendermint/core/types"
+)
+
+const (
+	MaxCommitWaitTimeSeconds = 10
+)
+
+type Confirmation struct {
+	BlockHash []byte
+	Event     txs.EventData
+	Exception error
+	Error     error
+}
+
+// NOTE [ben] Compiler check to ensure ErisNodeClient successfully implements
+// eris-db/client.NodeClient
+var _ NodeWebsocketClient = (*ErisNodeWebsocketClient)(nil)
+
+type ErisNodeWebsocketClient struct {
+	// TODO: assert no memory leak on closing with open websocket
+	tendermintWebsocket *rpcclient.WSClient
+}
+
+// Subscribe to an eventid
+func (erisNodeWebsocketClient *ErisNodeWebsocketClient) Subscribe(eventid string) error {
+	// TODO we can in the background listen to the subscription id and remember it to ease unsubscribing later.
+	return erisNodeWebsocketClient.tendermintWebsocket.Subscribe(eventid)
+}
+
+// Unsubscribe from an eventid
+func (erisNodeWebsocketClient *ErisNodeWebsocketClient) Unsubscribe(subscriptionId string) error {
+	return erisNodeWebsocketClient.tendermintWebsocket.Unsubscribe(subscriptionId)
+}
+
+// Returns a channel that will receive a confirmation with a result or the exception that
+// has been confirmed; or an error is returned and the confirmation channel is nil.
+func (erisNodeWebsocketClient *ErisNodeWebsocketClient) WaitForConfirmation(tx txs.Tx, chainId string, inputAddr []byte) (chan Confirmation, error) {
+	// check no errors are reported on the websocket 
+	if err := erisNodeWebsocketClient.assertNoErrors(); err != nil {
+		return nil, err
+	}
+
+	// Setup the confirmation channel to be returned
+	confirmationChannel := make(chan Confirmation, 1)
+	var latestBlockHash []byte
+
+	eid := txs.EventStringAccInput(inputAddr)
+
+	// Read the incoming events
+	go func() {
+		var err error
+		for {
+			resultBytes := <- erisNodeWebsocketClient.tendermintWebsocket.ResultsCh
+			result := new(ctypes.ErisDBResult)
+			if wire.ReadJSONPtr(result, resultBytes, &err); err != nil {
+				// keep calm and carry on
+				log.Errorf("eris-client - Failed to unmarshal json bytes for websocket event: %s", err)
+				continue
+			}
+			
+			event, ok := (*result).(*ctypes.ResultEvent)
+			if !ok {
+				// keep calm and carry on
+				log.Error("eris-client - Failed to cast to ResultEvent for websocket event")
+				continue
+			}
+			
+			blockData, ok := event.Data.(txs.EventDataNewBlock)
+			if ok {
+				latestBlockHash = blockData.Block.Hash()
+				log.WithFields(log.Fields{
+					"new block": blockData.Block,
+					"latest hash": latestBlockHash,
+				}).Debug("Registered new block")
+				continue
+			}
+			
+			// we don't accept events unless they came after a new block (ie. in)
+			if latestBlockHash == nil {
+				continue
+			}
+
+			if event.Event != eid {
+				log.Warnf("Received unsolicited event! Got %s, expected %s\n", event.Event, eid)
+				continue
+			}
+
+			data, ok := event.Data.(txs.EventDataTx)
+			if !ok {
```

Comment:
Here I think it *should* only be an EventDataTx for the event string formed from `InputAddr`; so this is stronger an error case than above I think